system/jlib/jptree.cpp

```diff
@@ -9481,3 +9542,341 @@ void copyPropIfMissing(IPropertyTree & target, const char * targetName, IPropert
             target.setProp(targetName, source.queryProp(sourceName));
     }
 }
+
+inline void checkEndHexSequence(StringBuffer &s, bool &hexsequence)
+{
+    if (hexsequence)
+    {
+        //close current hex stream sequence
+        s.append('_');
+        hexsequence = false;
+    }
+}
+inline bool isValidPTreeNameChar(const char *ch, unsigned remaining, unsigned &chlen, bool isatt, bool first)
+{
+    chlen = 0;
+    if (!ch || !*ch || !remaining)
+        return false;
+    chlen = utf8CharLen((const unsigned char *)ch, remaining);
+    if (0==chlen)
+    {
+        chlen=1;
+        return false;
+    }
+    if (chlen>1)
+        return !isatt;  //attributes do not seem to currently support multibyte characters in their names
+    if (first)
+        return isValidXPathStartChr(*ch);
+    return isValidXPathChr(*ch);
+}
+
+const char *findFirstInvalidPTreeNameChar(const char *ch, unsigned remaining)
+{
+    if (!ch)
+        return nullptr;
+    if (remaining==0 || *ch=='\0') //empty name still needs to trigger a special encoding
+        return ch;
+    bool isatt = ('@'==*ch);
+    if (isatt)
```

Comment:
avoid double condition? e.g.
```
     bool isAtt;
     if ('@' == *ch)
     {
         isAtt = true;
         ch++;
         remaining--;
     }
     else
          isAtt = false;
```

compiler may optimize to avoid double test anyway.